### ISSUE TITLE
torch.compile()/torch.export() compatibility ?

### ISSUE THREAD

**Issue by borisfom** [latkins/trifast#1]
Hi Liam,
Great job on trifast!
Do you have any work in the pipeline to make trifast compatible with torch.compile()/torch.export()? 
Looks like it would be relatively straightforward to modify existing code to use Torch 2.4+ custom op register methods in order to achieve that.

Best,
-Boris.

**Comment by latkins** [comment]
Hey, thanks! I've just pushed a WIP branch I had that does this: [liam/triton_op](https://github.com/latkins/trifast/tree/liam/triton_op). It didn't seem possible to do this while using the cached autotune configs, however, so I wasn't sure about pushing ahead. Would be curious to hear your thoughts!

**Comment by borisfom** [comment]
Oh, that's nice! Did @triton_op force you to change autotune? 
I was thinking of replacing whole Function with @custom_op / register_autograd, but triton_op may be used by that I suppose. Something that would pass this test,  to be able to be exported and later wrapped  into onnx or TensorRT plugin: 
```
    from torch.library import opcheck
    opcheck(
        torch.ops.trifast.triangle_attention,
        [q, k, v, b, m],
        test_utils="ALL",
    )

**Comment by borisfom** [comment]
Have you tried [triton-dejavu](https://github.com/IBM/triton-dejavu) for serialization ?

**Comment by latkins** [comment]
> Have you tried [triton-dejavu](https://github.com/IBM/triton-dejavu) for serialization ?

I wanted to use this, but it is only installable from source.

> Oh, that's nice! Did @triton_op force you to change autotune?

If you want the tracing to work, you can only pass `key`/`configs` to `autotune`, which is what tripped me up. Thinking about it, there is actually probably a way around this.

>  You may still want to revisit forward/backward op boundaries and use register_autograd instead of Function.

What is the advantage of doing this?

**Comment by borisfom** [comment]
> I wanted to use this, but it is only installable from source.

What's wrong with that ? You can have pip dependency that installs from source (specific commit).

**Comment by borisfom** [comment]
> What is the advantage of doing this?

Many. That's what Perplexity says : 
```
Advantages of `register_autograd`
	1.	Backend-Specific Support:
	•	`register_autograd` allows registering custom autograd kernels for specific backends (e.g., CUDA), enabling finer control over operations tailored to hardware.
	2.	Enhanced Integration with PyTorch Features:
	•	It works seamlessly with fake tensors used for flop counting or other analysis tools, which might not function well with traditional `autograd.Function`.
	•	It supports operator fusion in `torch.compile`, allowing operations like reshaping to merge into custom operators for better optimization.
	3.	Simplified Annotation for Mixed Precision:
	•	With `torch.library`, you can annotate support for features like automatic mixed precision (`torch.amp`) more effectively compared to the manual approach in `autograd.Function`.
	4.	Future-Proofing:
	•	As PyTorch evolves, `torch.library.register_autograd` is positioned as the preferred method for implementing custom operations, especially for workflows involving Triton or other advanced integrations.
```
It also makes difference for onnx.export etc. And is overall more advanced, preferred API moving forward.

**Comment by borisfom** [comment]
Regarding op boundaries : I suggest leaving only tensors as inputs, and move calculation of sm_scale, bh, n, h, dim (which are trivial) into the op: 
```
@triton_op("trifast::triangle_attention_forward", mutates_args={})
def fwd(
    q: torch.Tensor,
    k: torch.Tensor,
    v: torch.Tensor,
    b: torch.Tensor,
    mask: torch.Tensor,
    sm_scale: float,
    bh: int,
    n: int,
    h: int,
    dim: int,
) -> tuple[torch.Tensor, torch.Tensor]:

**Comment by borisfom** [comment]
And for _backward operation: same, plus you want to allocate return values in in instead of mutating arguments. There is also convention on what backward method should return, this doc may be a good start : https://docs.google.com/document/d/1_W62p8WJOQQUzPsJYa7s701JXt0qf2OfLub2sbkHOaU/edit?tab=t.0#heading=h.ptttacy8y1u9

**Comment by latkins** [comment]
Ah, that is a useful resource. I will have a read and move to this approach.

**Comment by latkins** [comment]
Have pushed an update to the above branch that passes opcheck. I'll fix the issue with autotune cache location, then push a new version.

**Comment by borisfom** [comment]
Thanks, the update looks great!

**Comment by latkins** [comment]
This should be done in now.
